src/CoreBundle/private/library/classes/TCMSTableEditor/TCMSTableFieldWriter.class.php

```diff
@@ -120,66 +136,118 @@ public function Save(&$postData, $bDataIsInSQLForm = false)
-        $this->_oField->sTableName = $this->_oParentRecord->sqlData['name'];
-        $this->_oField->recordId = $this->sId;
-        $this->_oField->oDefinition = &$this->oTable;
-        if ($bTypeChanged) {
-            $this->_oField->ChangeFieldTypePostHook();
+        $oldField->RemoveFieldIndex();
+
+        $this->oTable->sqlData['cms_field_type_id'] = $newTypeId;
+        /**
+         * @var $newField TCMSField
+         */
+        $newField = &$this->oTable->GetFieldObject();
+
+        $newField->name = $newName;
+        $newField->sTableName = $this->_oParentRecord->sqlData['name'];
+        $newField->recordId = $this->sId;
+        $newField->oDefinition = &$this->oTable;
+        if (true === $bTypeChanged) {
+            $newField->ChangeFieldTypePostHook();
         }
-        $this->_oField->ChangeFieldDefinition($oldName, $postData['name'], $postData);
-        $aOldData = $this->oTable->sqlData;
-        parent::Save($postData);
-        $this->oTable->Load($this->oTable->id);
-        if (!$bTargetTableChangeForMLTField && $this->_oField->AllowCreateRelatedTablesAfterFieldSave($aOldData, $currentFieldTypeRow, $newFieldTypeRow)) {
-            $this->_oField->CreateRelatedTables();
+
+        $oldFieldIsVirtual = $oldField->oDefinition->isVirtualField();
+        $newFieldIsVirtual = $newField->oDefinition->isVirtualField();
+
+        if ($oldFieldIsVirtual === $newFieldIsVirtual) {
+            $newField->ChangeFieldDefinition($this->oldData['name'], $newName, $postData);
+        } else {
+            if (false === $newFieldIsVirtual) {
+                $newField->CreateFieldDefinition(); // is missing: old one was virtual
+            } else {
+                $oldField->DeleteFieldDefinition();
+            }
         }
-        $this->_oField->CreateFieldIndex();
 
-        $this->_oField->oDefinition->UpdateFieldTranslationKeys($aOriginalData);
-        $this->getAutoclassesCacheWarmer()->updateTableById($this->oTable->sqlData['cms_tbl_conf_id']);
+        // NOTE here was Save() before the reorganization - hence the method names
+
+        if (false === $this->isChangeFromTable && true === $newField->AllowCreateRelatedTablesAfterFieldSave($this->oldData, $currentFieldTypeRow, $newFieldTypeRow)) {
+            $newField->CreateRelatedTables();
+        }
+        $newField->CreateFieldIndex();
+
+        $newField->oDefinition->UpdateFieldTranslationKeys($this->oldData);
+
+        return $newField;
+    }
+
+    /**
+     * @param string $typeId
+     *
+     * @return array|bool
+     *
+     * @throws DBALException
+     */
+    private function getFieldTypeDefinition(string $typeId)
+    {
+        $databaseConnection = $this->getDatabaseConnection();
+
+        $query = 'SELECT * FROM `cms_field_type` WHERE `id` = :typeId';
 
-        return $this->GetObjectShortInfo($postData);
+        return $databaseConnection->fetchAssoc($query, ['typeId' => $typeId]);
```

Comment:
Can we return null instead of false in case no rows are found? This would be consistent with other code from recent years.

Also the callers of this method don't expect false or null yet.

Reply:
ok: null returned

However this is a rather basic system error (missing type). For example subclasses will work with this data.
Should this then be some sort of RuntimeException?

Reply:
Maybe a `\LogicException`? Or what do you suggest (I'd like to avoid both `\Exception` and `\RuntimeException` as they don't express anything)?